br/pkg/restore/log_client/id_map.go

```diff
@@ -0,0 +1,212 @@
+	if checkpointStorage := rc.tryGetCheckpointStorage(logCheckpointMetaManager); checkpointStorage != nil {
+		log.Info("checkpoint storage is specified, load pitr id map from the checkpoint storage.")
+		if err := rc.saveIDMap2Storage(ctx, checkpointStorage, dbmaps); err != nil {
+			return errors.Trace(err)
+		}
+	} else if rc.pitrIDMapTableExists() {
+		if err := rc.saveIDMap2Table(ctx, dbmaps); err != nil {
+			return errors.Trace(err)
+		}
+	} else {
+		log.Info("the table mysql.tidb_pitr_id_map does not exist, maybe the cluster version is old.")
+		if err := rc.saveIDMap2Storage(ctx, rc.storage, dbmaps); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	if rc.useCheckpoint {
+		log.Info("save checkpoint task info with InLogRestoreAndIdMapPersist status")
+		if err := logCheckpointMetaManager.SaveCheckpointProgress(ctx, &checkpoint.CheckpointProgress{
+			Progress: checkpoint.InLogRestoreAndIdMapPersisted,
+		}); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (rc *LogClient) saveIDMap2Storage(
+	ctx context.Context,
+	storage storage.ExternalStorage,
+	dbMaps []*backuppb.PitrDBMap,
+) error {
+	clusterID := rc.GetClusterID(ctx)
+	metaFileName := pitrIDMapsFilename(clusterID, rc.restoreTS)
+	metaWriter := metautil.NewMetaWriter(storage, metautil.MetaFileSize, false, metaFileName, nil)
+	metaWriter.Update(func(m *backuppb.BackupMeta) {
+		m.ClusterId = clusterID
+		m.DbMaps = dbMaps
+	})
+	return metaWriter.FlushBackupMeta(ctx)
+}
+
+func (rc *LogClient) saveIDMap2Table(ctx context.Context, dbMaps []*backuppb.PitrDBMap) error {
+	backupmeta := &backuppb.BackupMeta{DbMaps: dbMaps}
+	data, err := proto.Marshal(backupmeta)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// clean the dirty id map at first
+	err = rc.unsafeSession.ExecuteInternal(ctx, "DELETE FROM mysql.tidb_pitr_id_map WHERE restored_ts = %? and upstream_cluster_id = %?;", rc.restoreTS, rc.upstreamClusterID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	replacePitrIDMapSQL := "REPLACE INTO mysql.tidb_pitr_id_map (restored_ts, upstream_cluster_id, segment_id, id_map) VALUES (%?, %?, %?, %?);"
+	for startIdx, segmentId := 0, 0; startIdx < len(data); segmentId += 1 {
+		endIdx := startIdx + PITRIdMapBlockSize
+		if endIdx > len(data) {
+			endIdx = len(data)
+		}
+		err := rc.unsafeSession.ExecuteInternal(ctx, replacePitrIDMapSQL, rc.restoreTS, rc.upstreamClusterID, segmentId, data[startIdx:endIdx])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		startIdx = endIdx
+	}
+	return nil
+}
+
+func (rc *LogClient) loadSchemasMap(
+	ctx context.Context,
+	restoredTS uint64,
+	logCheckpointMetaManager checkpoint.LogMetaManagerT,
+) ([]*backuppb.PitrDBMap, error) {
+	if checkpointStorage := rc.tryGetCheckpointStorage(logCheckpointMetaManager); checkpointStorage != nil {
+		log.Info("checkpoint storage is specified, load pitr id map from the checkpoint storage.")
+		dbMaps, err := rc.loadSchemasMapFromStorage(ctx, checkpointStorage, restoredTS)
```

Comment:
should fallback here when got error?

Reply:
Which one is better, if specify checkpoint storage but got error, fallback to table or report error?

What if user first time restore without checkpoint storage, and try start from checkpoint with a new parameter checkpoint storage?

Reply:
After talk with @Leavrth , we could do the check in another PR, to avoid a kind of ABA problem